libbeat/common/jsontransform/jsonhelper.go

```diff
@@ -35,11 +35,11 @@ var (
 	// ErrInvalidTimestamp is returned when parsing of a @timestamp field fails.
 	// Supported formats: ISO8601, RFC3339
 	ErrInvalidTimestamp = errors.New("failed to parse @timestamp, unknown format")
+	logger              = logp.NewLogger("jsonhelper")
```

Comment:
Does this logger still work? This become and order of initialization problem. The logp package is not yet initialized at this point so you get a no-op logger returned.

Reply:
When tested in the unit/benchmark tests, it was properly initialized. 
I can move this into a `func init()` on this package or see if we propagate it as part of the function parameters from the caller.

Reply:
> it was properly initialized.

It will be initialized, but I would argue not properly. Moving it to an `init()` will not fix the issue since that still happens before the `logp` package receives its configuration. The [godoc](https://github.com/elastic/elastic-agent-libs/blob/16bf3dfcb3655978896c7d081b1b4e3dbf967046/logp/logger.go#L44-L46) for `NewLogger` calls this out:

> // NewLogger returns a new Logger labeled with the name of the selector. This
> // should never be used from any global contexts, otherwise you will receive a
> // no-op Logger. This is because the logp package needs to be initialized first.

If you apply this diff and run the test again, nothing gets logged.

`go test -v -run TestWriteJSONKeys`

```diff
diff --git a/libbeat/common/jsontransform/jsonhelper.go b/libbeat/common/jsontransform/jsonhelper.go
index 935eac90a7..c1e9a5d05b 100644
--- a/libbeat/common/jsontransform/jsonhelper.go
+++ b/libbeat/common/jsontransform/jsonhelper.go
@@ -40,6 +40,7 @@ var (
 
 // WriteJSONKeys writes the json keys to the given event based on the overwriteKeys option and the addErrKey
 func WriteJSONKeys(event *beat.Event, keys map[string]interface{}, expandKeys, overwriteKeys, addErrKey bool) {
+	logger.Error("NEVER SHOWN")
 	if expandKeys {
 		if err := expandFields(keys); err != nil {
 			logger.Errorf("JSON: failed to expand fields: %s", err)
diff --git a/libbeat/common/jsontransform/jsonhelper_test.go b/libbeat/common/jsontransform/jsonhelper_test.go
index d45ca5c3a3..9f43c1dbb3 100644
--- a/libbeat/common/jsontransform/jsonhelper_test.go
+++ b/libbeat/common/jsontransform/jsonhelper_test.go
@@ -24,10 +24,13 @@ import (
 	"github.com/stretchr/testify/require"
 
 	"github.com/elastic/beats/v7/libbeat/beat"
+	"github.com/elastic/elastic-agent-libs/logp"
 	"github.com/elastic/elastic-agent-libs/mapstr"
 )
 
 func TestWriteJSONKeys(t *testing.T) {
+	logp.TestingSetup()
+
 	now := time.Now()
 	now = now.Round(time.Second)
```

------

## Solutions

1. One solution would be to defer initialization of the logger until the time of first use. Like

```
var (
	pkgLogger         *logp.Logger
	pkgLoggerInitOnce sync.Once
)

func logger() *logp.Logger {
	pkgLoggerInitOnce.Do(func() {
		pkgLogger = logp.NewLogger("jsontransform")
	})
	return pkgLogger
}

func WriteJSONKeys(...) {
    logger().Errorf("JSON: failed to expand fields: %s", err)
}
```

2. Remove logging from this function. I kind of find it annoying that it does any logging. There is already error handling present and that is via the `addErrKey` feature. As a user, if I care about errors then I should be using `addErrKey==true`.
\
This highlights another inefficiency. When `addErrKey==false` the code is always executing `createJSONError(err.Error())` which then goes completely unused.